[CS-3004] Stores hub auth token

oidoug opened the pull request:
<!-- This is a pull request. Please make sure any applicable bullet points have been covered. -->

### Description

Saves Hub Auth Token
- In hub-service.getHubAuthToken, tries to load a token from local storage.
- If none is present, generate a new one with authAPI.authenticate and store it locally.
- If for some reason the saved token gets rejected by the hub API, the stored token pointing to stored key 'hubToken-[NETWORK]' is deleted.
- Some minor improvements were made to common storage calls `saveLocal` and `getLocal` to allow for using of `ids` alongside keys, so deleting entries with a shared key is a lot easier.
 
Refactored calls to HDProvider.reset to single place in getHubAuthToken.

- [x] Completes #(CS-3004)

### Checklist

- [x] All UI changes have been tested on a small device

### Screenshots

https://user-images.githubusercontent.com/129619/150995647-d29b93a7-f696-4a75-94d3-c27b11278fb2.mp4

Commit:
Adds new Notifications Preferences
- Notifications section to Settings screen;
- Service update on hub-service for getting and updating notif. options;
- Hook useUpdateNotificationPreferences for handling authToken and options state.

Commit:
Fix refactor mistake: wrong params order.

Commit:
Fix for refactor mistake: removing unused OptionsList.

Commit:
Saves Hub Auth Token
- in hub-service.getHubAuthToken, tries to load a token from local storage
- if none present, generates new with authAPI.authenticate and stores locally.
Refactored calls to HDProvider.reset to single place in getHubAuthToken.

oidoug commented:
Note: Just notice it's not yet handling if the saved token is invalidated. If a request errors on a token we should try to re-create it.

Commit:
Refactor, moves fcmToken fetching to service;
Adds better memos to NotificationsSections inner components.

Commit:
Merge branch 'add-notification-panel-cs-2569' into store-hub-auth-token-and-re-use-it-if-cs-3004

Commit:
Removed network from key

Commit:
WIP: Removing auth token when request fails with 401

Commit:
Removing auth token when request fails with 401;
Not able to retry connection since token is also handled outside of the hub-service scope - we need to refactort that.

Commit:
Merge branch 'develop' into store-hub-auth-token-and-re-use-it-if-cs-3004

Commit:
Checking for null on axios interceptor;
Updated package react-native-flipper.

topwebtek7 commented on cardstack/src/services/hub-service.ts:
```diff
@@ -355,3 +405,19 @@ export const getWyrePrice = async (
   }
 };
+
+axios?.interceptors?.response?.use?.(undefined, async (error: AxiosError) => {
+  if (error?.response?.status !== 401) {
+    return Promise.reject(error);
+  }
+  // Got API error 401, auth token should be refreshed.
+
+  // Remove token from local storage
+  const network: Network = await getNetwork();
+  await removeLocal(hubTokenStorageKey(network));
```
Can we filter only Hub Api calls here? I think we're using Axios for other api calls as well like Wyre, OpenSea, etc..

oidoug replied:
Oh, good catch. I think this interceptor only acts on this file's context, but just to be sure I'll add a specific instance of axios for the hub calls.

topwebtek7 commented on package.json:
```diff
@@ -194,7 +194,7 @@
     "react-native-extra-dimensions-android": "1.2.2",
     "react-native-fast-image": "^8.1.5",
-    "react-native-flipper": "^0.114.0",
+    "react-native-flipper": "^0.130.0",
```
I think will need to add Pod lock file change as well if upgraded deps

oidoug replied:
I don't think this upgrade is needed now, reverting.

topwebtek7 commented on cardstack/src/services/hub-service.ts:
```diff
@@ -36,6 +43,38 @@ const axiosConfig = (authToken: string) => {
+  return authToken;
+};
+
+const storeHubAuthToken = async (
+  tokenStorageKey: string,
+  authToken: string,
+  walletAddress: string
+) => {
+  // expires in a day.
+  await saveLocal(
+    tokenStorageKey,
+    { data: { authToken } },
+    1000 * 3600 * 24,
+    undefined,
+    walletAddress
+  );
+};
```
hmm In my opinion, it's not a good idea use `walletAddress`  as `version` parameter of `saveLocal` function.. I think version parameter is meant to be used when need to change specific asyncStorage key's behavior/structure at some point of app updates.
Instead, we can generate keys with address, network value, like as `getKey` function does

```
export const getKey = (prefix, accountAddress, network) =>
  `${prefix}-${toLower(accountAddress)}-${toLower(network)}`;
```

oidoug replied:
I updated the `saveLocal` signature, so the `walletAddress` in this case is actually an ID that will be used alongside the key to store the token. Version parameter is set to `undefined` (so will use the default value).

I added this ID (supported by rn-storage) so it's easier to remove a group of entries by a shared key - that in this case uses the current network as key and wallletaddress as id.

topwebtek7 commented on cardstack/src/services/hub-service.ts:
```diff
@@ -64,9 +116,11 @@ export const getHubAuthToken = async (
-    // load wallet address when not provided as an argument(this keychain access does not require passcode/biometric auth)
-    const address = walletAddress || (await loadAddress()) || '';
     const authToken = await authAPI.authenticate({ from: address });
+    await HDProvider.reset();
```
The idea behind of `getHubAuthToken` didn't have `await HDProvider.reset();` inside of it was - just to improve performance of getHubAuthToken a little bit, as we can reset HDProvider(resetting HDProvider's async action) even after authToken is returned, just FYI.. but seems it's really tiny delay so looks ok

oidoug replied:
Now getHubAuthToken pull the auth token from storage instead of requesting it from HDProvider.
That has drastically speedup the token pullup process, so leaving reset here isn't an issue anymore.

topwebtek7 reviewed (commented):
Looks good so far - left some comments

Commit:
Adds custom instance for Axios in Hub-Service;
Reverts rn-flipper to previous version (same as pod.lock one).

Commit:
Uses custom axios instance only for new untested functions.
Couldnt mock axios instance properly for jest to work, so old tests were failing with axios instance.

oidoug commented:
@topwebtek7 , I wasn't able to update the jest tests to work with axios's custom instance. Since we're probably going to refactor this to use RTK, I don't want to spend too much more time on this, so I just used the instance in the untested calls. Let me know if you are ok with this or if you would prefer it differently.

danibonilha commented on cardstack/src/services/hub-service.ts:
```diff
@@ -36,6 +43,40 @@ const axiosConfig = (authToken: string) => {
   };
 };
 
+export const hubapi = axios.create(axiosConfig(''));
```
`hubApi`

danibonilha commented on yarn.lock:
```diff
@@ -17647,10 +17647,10 @@ react-native-fast-image@^8.1.5:
+react-native-flipper@^0.130.0:
+  version "0.130.0"
+  resolved "https://registry.yarnpkg.com/react-native-flipper/-/react-native-flipper-0.130.0.tgz#91e907257000af514c4173c8f7b748943c08cf5c"
+  integrity sha512-F6UzGTRSy24tEFroQGxSfm6+7qL3LuV8RYG3O/91vGzUKX+IBZdRdEJWeuELk/6WBtFAUuWEE7mlIWe0ZwwrQQ==
```
How is updating flipper related to this task ? If we do need to update I think the package.json change is missing then.

oidoug replied:
It's not needed, I had issues and tried updating, but it wasn't related. Reverting...

danibonilha reviewed (approved):
LGTM, just some small nits, and regarding the tests we can skip for this instance since it’s temporary

Commit:
Renamed hubapi to hubApi;
Restored develop yarn.lock.

Commit:
Merge branch 'develop' into store-hub-auth-token-and-re-use-it-if-cs-3004
